File: scripts/check-header-init-method.py
#!/usr/bin/env python3
"""
Script to check header initialization method across all pages
בודק בפועל איזה שיטת איתחול header רץ בכל עמוד
"""

import os
import json
import time
import subprocess
from pathlib import Path
from collections import defaultdict

# Try to import selenium
try:
    from selenium import webdriver
    from selenium.webdriver.chrome.options import Options
    from selenium.webdriver.chrome.service import Service
    from selenium.webdriver.common.by import By
    from selenium.webdriver.support.ui import WebDriverWait
    from selenium.webdriver.support import expected_conditions as EC
    from selenium.common.exceptions import TimeoutException, WebDriverException
    SELENIUM_AVAILABLE = True
except ImportError:
    SELENIUM_AVAILABLE = False
    print("⚠️  Selenium not available. Will create manual check script instead.")

# Get project root
project_root = Path(__file__).parent.parent
trading_ui_dir = project_root / 'trading-ui'

# Auth pages to skip
AUTH_PAGES = ['login.html', 'register.html', 'forgot-password.html', 'reset-password.html']

def get_all_html_pages():
    """Get all HTML pages except auth pages and test pages"""
    pages = []
    for html_file in trading_ui_dir.glob('*.html'):
        filename = html_file.name
        
        # Skip auth pages
        if any(auth in filename for auth in AUTH_PAGES):
            continue
        
        # Skip test pages and smart pages
        if filename.startswith('test-') or filename.endswith('-smart.html'):
            continue
        
        # Skip mockups
        if 'mockup' in filename.lower():
            continue
        
        pages.append(html_file)
    
    return sorted(pages)

def check_with_selenium(page_path, base_url='http://localhost:8080'):
    """Check header initialization method using Selenium"""
    if not SELENIUM_AVAILABLE:
        return None
    
    try:
        # Setup Chrome options
        chrome_options = Options()
        chrome_options.add_argument('--headless')
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-dev-shm-usage')
        chrome_options.add_argument('--disable-gpu')
        chrome_options.add_argument('--window-size=1920,1080')
        
        # Create driver
        driver = webdriver.Chrome(options=chrome_options)
        
        try:
            # Navigate to page
            page_name = page_path.name
            url = f"{base_url}/{page_name}"
            print(f"  Loading: {url}")
            driver.get(url)
            
            # Wait for page to load (wait for console logs)
            time.sleep(5)  # Wait for initialization
            
            # Get console logs
            logs = driver.get_log('browser')
            
            # Check for initialization method in logs
            method = None
            for log in logs:
                message = log.get('message', '')
                if 'HEADER INIT' in message:
                    if 'PLANNED' in message:
                        method = 'planned'
                        break
                    elif 'FALLBACK' in message:
                        method = 'fallback'
                        break
            
            # Also check localStorage
            if not method:
                try:
                    init_logs = driver.execute_script("return localStorage.getItem('__headerInitLogs');")
                    if init_logs:
                        logs_data = json.loads(init_logs)
                        if logs_data:
                            last_log = logs_data[-1]
                            if last_log.get('page', '').endswith(page_name):
                                method = last_log.get('method', 'unknown')
                except:
                    pass
            
            return method
            
        finally:
            driver.quit()
            
    except Exception as e:
        print(f"  Error: {e}")
        return None

def create_manual_check_script(pages):
    """Create a script that can be run manually in browser console"""
    script_content = f"""
// Manual Header Init Method Checker
// Run this in browser console on each page

(function() {{
    'use strict';
    
    const pages = {json.dumps([p.name for p in pages], indent=2)};
    
    function checkCurrentPage() {{
        const pageName = window.location.pathname.split('/').pop();
        const initLogs = JSON.parse(localStorage.getItem('__headerInitLogs') || '[]');
        const pageLog = initLogs.find(log => log.page.endsWith(pageName));
        
        const result = {{
            page: pageName,
            method: window.__headerSystemInitMethod || pageLog?.method || 'unknown',
            headerInitialized: !!(window.headerSystem && window.headerSystem.isInitialized),
            timestamp: new Date().toISOString()
        }};
        
        console.log('📊 HEADER INIT RESULT:', JSON.stringify(result, null, 2));
        return result;
    }}
    
    // Check current page
    const result = checkCurrentPage();
    
    // Store result
    if (!window.__headerInitResults) {{
        window.__headerInitResults = [];
    }}
    window.__headerInitResults.push(result);
    
    console.log('✅ Result stored in window.__headerInitResults');
    console.log('📋 To see all results: console.log(JSON.stringify(window.__headerInitResults, null, 2))');
    
    return result;
}})();
"""
    
    script_file = project_root / 'scripts' / 'manual-header-check.js'
    with open(script_file, 'w', encoding='utf-8') as f:
        f.write(script_content)
    
    return script_file

def main():
    print("🔍 Checking header initialization method across all pages...")
    print("=" * 70)
    
    pages = get_all_html_pages()
    print(f"Found {len(pages)} pages to check\n")
    
    results = {
        'planned': [],
        'fallback': [],
        'unknown': [],
        'errors': []
    }
    
    # Check if server is running
    try:
        import urllib.request
        urllib.request.urlopen('http://localhost:8080', timeout=2)
        server_running = True
        print("✅ Server is running on http://localhost:8080\n")
    except:
        server_running = False
        print("⚠️  Server is not running. Cannot check pages automatically.")
        print("   Please start the server with: ./start_server.sh\n")
    
    if server_running and SELENIUM_AVAILABLE:
        print("Using Selenium to check pages...\n")
        for page in pages:
            page_name = page.name
            print(f"Checking {page_name}...", end=' ')
            method = check_with_selenium(page)
            
            if method:
                results[method].append(page_name)
                print(f"✅ {method.upper()}")
            else:
                results['unknown'].append(page_name)
                print("❓ UNKNOWN")
    else:
        # Create manual check script
        print("Creating manual check script...\n")
        script_file = create_manual_check_script(pages)
        print(f"✅ Created manual check script: {script_file}")
        print("\nTo use it:")
        print("1. Open each page in your browser")
        print("2. Open browser console (F12)")
        print("3. Copy and paste the content of the script")
        print("4. Check the result in console\n")
        
        # For now, just show which pages have init-system
        print("Based on file analysis:\n")
        for page in pages:
            page_name = page.name
            has_init = 'package-manifest.js' in page.read_text(encoding='utf-8') or \
                      'page-initialization-configs.js' in page.read_text(encoding='utf-8')
            
            if has_init:
                results['planned'].append(page_name)
                print(f"✅ {page_name:40} - Has init-system (likely PLANNED)")
            else:
                results['fallback'].append(page_name)
                print(f"🔄 {page_name:40} - No init-system (will use FALLBACK)")
    
    print("\n" + "=" * 70)
    print("📊 SUMMARY:")
    print("=" * 70)
    print(f"✅ PLANNED method: {len(results['planned'])} pages")
    print(f"🔄 FALLBACK method: {len(results['fallback'])} pages")
    print(f"❓ UNKNOWN: {len(results['unknown'])} pages")
    if results['errors']:
        print(f"❌ ERRORS: {len(results['errors'])} pages")
    
    # Save results
    results_file = project_root / 'documentation' / '05-REPORTS' / 'header-init-actual-results.json'
    results_file.parent.mkdir(parents=True, exist_ok=True)
    
    with open(results_file, 'w', encoding='utf-8') as f:
        json.dump(results, f, indent=2, ensure_ascii=False)
    
    print(f"\n💾 Results saved to: {results_file}")
    
    return results

if __name__ == '__main__':
    main()

File: scripts/check-header-init-method.js
#!/usr/bin/env node
/**
 * Script to check header initialization method across all pages
 * בודק בפועל איזה שיטת איתחול header רץ בכל עמוד
 */

const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..');
const tradingUiDir = path.join(projectRoot, 'trading-ui');

// Auth pages to skip
const AUTH_PAGES = ['login.html', 'register.html', 'forgot_password.html', 'reset_password.html'];

function getAllHtmlPages() {
  const pages = [];
  const files = fs.readdirSync(tradingUiDir);
  
  for (const file of files) {
    if (!file.endsWith('.html')) continue;
    
    // Skip auth pages
    if (AUTH_PAGES.some(auth => file.includes(auth))) continue;
    
    // Skip test pages and smart pages
    if (file.startsWith('test-') || file.endsWith('-smart.html')) continue;
    
    // Skip mockups
    if (file.toLowerCase().includes('mockup')) continue;
    
    pages.push(file);
  }
  
  return pages.sort();
}

async function checkPage(browser, pageName, baseUrl = 'http://localhost:8080') {
  const page = await browser.newPage();
  
  try {
    // Listen to console messages and capture HEADER INIT logs
    const consoleMessages = [];
    let headerInitMethod = null;
    
    page.on('console', msg => {
      const text = msg.text();
      consoleMessages.push(text);
      
      // Check for HEADER INIT messages
      if (text.includes('[HEADER INIT]')) {
        if (text.includes('PLANNED')) {
          headerInitMethod = 'planned';
        } else if (text.includes('FALLBACK')) {
          headerInitMethod = 'fallback';
        }
      }
    });
    
    // Navigate to page
    const url = `${baseUrl}/${pageName}`;
    console.log(`  Loading: ${url}`);
    
    await page.goto(url, { waitUntil: 'networkidle2', timeout: 30000 });
    
    // Wait longer for initialization (10 seconds to ensure everything loads and logs are written)
    await new Promise(resolve => setTimeout(resolve, 10000));
    
    // Check window.__headerSystemInitMethod first (most reliable)
    let method = headerInitMethod; // Use value captured from console listener
    
    if (!method) {
      try {
        const debugInfo = await page.evaluate(() => {
          return {
            headerSystemInitMethod: window.__headerSystemInitMethod || null,
            headerSystemExists: typeof window.headerSystem !== 'undefined',
            headerSystemInitialized: !!(window.headerSystem && window.headerSystem.isInitialized),
            unifiedAppExists: typeof window.UnifiedAppInitializer !== 'undefined',
            unifiedAppInitialized: window.globalInitializationState?.unifiedAppInitialized || false,
            localStorageLogs: localStorage.getItem('__headerInitLogs')
          };
        });
        
        method = debugInfo.headerSystemInitMethod;
        
        // If method not set but header is initialized, infer from context
        if (!method && debugInfo.headerSystemInitialized) {
          if (debugInfo.unifiedAppExists && debugInfo.unifiedAppInitialized) {
            method = 'planned';
          } else if (debugInfo.headerSystemInitialized) {
            method = 'fallback';
          }
        }
        
        // Check localStorage as backup
        if (!method && debugInfo.localStorageLogs) {
          try {
            const logs = JSON.parse(debugInfo.localStorageLogs);
            const pageLog = logs.find(log => {
              const logPage = log.page || '';
              return logPage.endsWith(pageName) || logPage.includes(pageName);
            });
            if (pageLog && pageLog.method) {
              method = pageLog.method;
            }
          } catch (e) {
            // Ignore parse errors
          }
        }
      } catch (e) {
        // Ignore evaluation errors
      }
    }
    
    // Last resort: check console messages we captured
    if (!method) {
      for (const msg of consoleMessages) {
        if (msg.includes('[HEADER INIT]')) {
          if (msg.includes('PLANNED')) {
            method = 'planned';
            break;
          } else if (msg.includes('FALLBACK')) {
            method = 'fallback';
            break;
          }
        }
      }
    }
    
    // Last resort: check if header is initialized and infer method
    if (!method) {
      try {
        const headerInfo = await page.evaluate(() => {
          return {
            headerInitialized: !!(window.headerSystem && window.headerSystem.isInitialized),
            hasUnifiedApp: typeof window.UnifiedAppInitializer !== 'undefined',
            unifiedAppInitialized: window.globalInitializationState?.unifiedAppInitialized || false
          };
        });
        
        // If header is initialized and UnifiedAppInitializer exists, likely planned
        if (headerInfo.headerInitialized && headerInfo.hasUnifiedApp) {
          method = 'planned';
        } else if (headerInfo.headerInitialized) {
          method = 'fallback';
        }
      } catch (e) {
        // Ignore
      }
    }
    
    return method || 'unknown';
    
  } catch (error) {
    console.error(`  Error checking ${pageName}:`, error.message);
    return 'error';
  } finally {
    await page.close();
  }
}

async function main() {
  console.log('🔍 Checking header initialization method across all pages...');
  console.log('='.repeat(70));
  
  const pages = getAllHtmlPages();
  console.log(`Found ${pages.length} pages to check\n`);
  
  // Check if server is running
  try {
    const http = require('http');
    await new Promise((resolve, reject) => {
      const req = http.get('http://localhost:8080', (res) => {
        resolve();
      });
      req.on('error', reject);
      req.setTimeout(2000, () => reject(new Error('Timeout')));
    });
    console.log('✅ Server is running on http://localhost:8080\n');
  } catch (error) {
    console.error('❌ Server is not running. Please start it with: ./start_server.sh\n');
    process.exit(1);
  }
  
  const results = {
    planned: [],
    fallback: [],
    unknown: [],
    errors: []
  };
  
  console.log('Launching browser...\n');
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  try {
    for (let i = 0; i < pages.length; i++) {
      const pageName = pages[i];
      process.stdout.write(`[${i + 1}/${pages.length}] Checking ${pageName}... `);
      
      const method = await checkPage(browser, pageName);
      
      results[method].push(pageName);
      
      const emoji = method === 'planned' ? '✅' : method === 'fallback' ? '🔄' : method === 'error' ? '❌' : '❓';
      console.log(`${emoji} ${method.toUpperCase()}`);
    }
  } finally {
    await browser.close();
  }
  
  console.log('\n' + '='.repeat(70));
  console.log('📊 SUMMARY:');
  console.log('='.repeat(70));
  console.log(`✅ PLANNED method: ${results.planned.length} pages`);
  console.log(`🔄 FALLBACK method: ${results.fallback.length} pages`);
  console.log(`❓ UNKNOWN: ${results.unknown.length} pages`);
  if (results.errors.length > 0) {
    console.log(`❌ ERRORS: ${results.errors.length} pages`);
  }
  
  // Save results
  const resultsFile = path.join(projectRoot, 'documentation', '05-REPORTS', 'header-init-actual-results.json');
  fs.mkdirSync(path.dirname(resultsFile), { recursive: true });
  fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2), 'utf-8');
  
  console.log(`\n💾 Results saved to: ${resultsFile}`);
  
  // Create markdown report
  const reportFile = path.join(projectRoot, 'documentation', '05-REPORTS', 'HEADER_INIT_ACTUAL_RESULTS.md');
  let report = `# Header Initialization Method - Actual Results\n\n`;
  report += `**Generated:** ${new Date().toISOString()}\n\n`;
  report += `## Summary\n\n`;
  report += `- **Planned Method:** ${results.planned.length} pages\n`;
  report += `- **Fallback Method:** ${results.fallback.length} pages\n`;
  report += `- **Unknown:** ${results.unknown.length} pages\n`;
  if (results.errors.length > 0) {
    report += `- **Errors:** ${results.errors.length} pages\n`;
  }
  report += `\n## Pages with Planned Method\n\n`;
  for (const page of results.planned.sort()) {
    report += `- \`${page}\`\n`;
  }
  if (results.fallback.length > 0) {
    report += `\n## Pages with Fallback Method\n\n`;
    for (const page of results.fallback.sort()) {
      report += `- \`${page}\`\n`;
    }
  }
  if (results.unknown.length > 0) {
    report += `\n## Pages with Unknown Method\n\n`;
    for (const page of results.unknown.sort()) {
      report += `- \`${page}\`\n`;
    }
  }
  if (results.errors.length > 0) {
    report += `\n## Pages with Errors\n\n`;
    for (const page of results.errors.sort()) {
      report += `- \`${page}\`\n`;
    }
  }
  
  fs.writeFileSync(reportFile, report, 'utf-8');
  console.log(`📄 Report saved to: ${reportFile}`);
  
  return results;
}

if (require.main === module) {
  main().catch(error => {
    console.error('Error:', error);
    process.exit(1);
  });
}

module.exports = { main, checkPage, getAllHtmlPages };

